src/ApplicationStatus.hpp

```diff
@@ -0,0 +1,150 @@
+            ///        node-local MPI rank; a Linux parent process ID
+            ///        for an application; or some other concept of a
+            ///        process with its own memory.  This ID should be
+            ///        self-consistent and unique within a job.  There
+            ///        will be one Profile object per process on the
+            ///        application side, and one ApplicationRecordLog
+            ///        per process on each side of the shared memory.
+            virtual void set_process(const std::set<int> &cpu_idx, int process) = 0;
+            /// @brief Get the process ID for the process the CPU is
+            ///        currently assigned to.
+            virtual int get_process(int cpu_idx) const = 0;
+            /// @brief Get the process IDs for every CPU.
+            virtual std::vector<int> get_process(void) const = 0;
+            /// @brief Create an ApplicationStatus object using the
+            ///        given SharedMemory.  The caller is responsible
+            ///        for calling `buffer_size()` when creating the
+            ///        shared memory, or attaching to an existing
+            ///        shared memory region before passing the object
+            ///        to this method.
+            static std::unique_ptr<ApplicationStatus> make_unique(const PlatformTopo& topo,
+                                                                  std::shared_ptr<SharedMemory> shmem);
+            /// @brief Return the required size of the shared memory
+            ///        region used by the ApplicationStatus for the
+            ///        given number of CPUs.
+            static size_t buffer_size(int num_cpu);
+
+        protected:
+            static constexpr size_t M_STATUS_SIZE = geopm::hardware_destructive_interference_size;
+    };
+
+    class ApplicationStatusImp : public ApplicationStatus
+    {
+        public:
+            ApplicationStatusImp(const PlatformTopo& topo,
+                                 std::shared_ptr<SharedMemory> shmem);
+            virtual ~ApplicationStatusImp() = default;
+            void set_hint(int cpu_idx, uint64_t hints) override;
+            uint64_t get_hint(int cpu_idx) const override;
+            std::vector<uint64_t> get_hint(void) const override;
+            void set_total_work_units(int cpu_idx, int work_units) override;
+            void increment_work_unit(int cpu_idx) override;
+            double get_work_progress(int cpu_idx) const override;
+            std::vector<double> get_work_progress(void) const override;
+            void set_process(const std::set<int> &cpu_idx, int process) override;
+            int get_process(int cpu_idx) const override;
+            std::vector<int> get_process(void) const override;
+        private:
+            // These fields must all be 32-bit int
+            struct m_app_status_s
+            {
+                uint32_t hints;
+                int32_t process; // can be negative
```

Comment:
not really important, but maybe process should come first, since it is set first?